File: mobile/src/services/api/Ingredients.ts
import { HttpMethod, callApi } from "../Server";
import { ingredientToJson, parseIngredients } from "../IngredientParser";
import { Ingredient } from "../../model/Ingredient";
import { parseNull } from "../NullParser";

export class IngredientsApi {
    public static addIngredient(newIngredient: Ingredient): Promise<number> {
        return callApi(
                "ingredients",
                HttpMethod.POST,
                parseInt,
                ingredientToJson(newIngredient));
    }

    public static getAllIngredients(): Promise<Ingredient[]> {
        return callApi(
                "ingredients",
                HttpMethod.GET,
                parseIngredients);
    }

    public static deleteIngredient(ingredientId: number): Promise<void> {
        return callApi(
                "ingredients",
                HttpMethod.DELETE,
                parseNull,
                JSON.stringify(ingredientId));
    }

    public static updateIngredient(updatedIngredient: Ingredient): Promise<void> {
        return callApi(
                "ingredients",
                HttpMethod.PUT,
                parseNull,
                ingredientToJson(updatedIngredient));
    }
} 

File: services/Recipes/Recipes/Interface/Ingredients.fs
namespace Interface

open Model
open DataAccess

module Ingredients =
    let add (ingredient: Ingredient) : Result<int, Error> =
        IngredientDataAccess.addIngredient ingredient
        |> Result.Ok

    let getAll () : Result<Ingredient[], Error> =
        IngredientDataAccess.getAllIngredients ()
        |> Result.Ok 

    let delete (ingredientId: int) : Result<unit, Error> =
        let anyRecipeUsesIngredient: bool = IngredientDataAccess.anyRecipeContainsIngredient ingredientId
        match anyRecipeUsesIngredient with
        | true -> Result.Error Error.IngredientIsUsedByARecipe
        | false ->
            IngredientNutritionDataAccess.deleteIngredientNutritionEntry ingredientId |> ignore
            IngredientDataAccess.deleteIngredient ingredientId |> ignore
            Result.Ok ()

    let update (newIngredient: Ingredient) : Result<unit, Error> =
        IngredientDataAccess.updateIngredient newIngredient |> ignore
        Result.Ok ()